### PR TITLE
Optimize PUT - Dont fetch info of not existing file

### PR DESCRIPTION
**Pull request by mrow4a**
"in case the file already exists/overwriting" ? In createFile() function which from composer is used only for new files, and there is specific one called updateFile() ?

-1 query plus Locking/Releasing because each getFileInfo needs to do that
selection_295

### CODE REVIEW

**Comment by PVince81** [comment]
Did you test ? Is `updateFile()` called for overwrites and not `createfile()` ? It is likely that you are right.

From what I remember, an overwrite still goes to `File::put()` though.

**Comment by PVince81** [comment]
make sure to test the following:
- [ ] old endpoint overwrite upload
- [ ] old endpoint chunked upload+overwrite
- [ ] new endpoint overwrite upload
- [ ] new endpoint chunked upload-overwrite

there is a slight chance that the chunked mode will go through `createFile()` even for overwrites... (yes, messy stuff here...)

**Comment by PVince81** [comment]
@mrow4a any update on my comments ?

**Comment by PVince81** [comment]
@mrow4a please rebase. On master I've added chunking + overwrite tests so the integration tests will take care of verifying your change.

If tests pass, then 👍

**Comment by PVince81** [comment]
let me rebase that for you

**Comment by PVince81** [comment]
```
Resuming build at Tue Jul 04 13:24:01 CEST 2017 after Jenkins restart
Ready to run at Tue Jul 04 13:24:02 CEST 2017
13:24:02 Timeout set to expire in 1 hr 29 min
```

Rebased ⛏

**Comment by PVince81** [comment]
some apparently legit failures:
```
16:35:57 --- Failed scenarios:
16:35:57 
16:35:57     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/quota.feature:24
16:35:57     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/quota.feature:74
16:35:57     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/quota.feature:107
16:35:57     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/webdav-related-new-endpoint.feature:482
```

@mrow4a please have a look

**Comment by PVince81** [comment]
```
01:48:30 
01:48:30 --- Failed scenarios:
01:48:30 
01:48:30     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/quota.feature:24
01:48:30     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/quota.feature:74
01:48:30     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/quota.feature:107
01:48:30     /var/lib/jenkins/workspace/owncloud-core_core_PR-28021-LBIG7E7PMTL3SCH6VDBSM2DPVZILEZZV2SKQOI7W3QQLEFOBH4CQ/tests/integration/features/webdav-related-new-endpoint.feature:519
01:48:30 
```

**Comment by PVince81** [comment]
```

  | Test Result (2 failures  / -2)OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\UploadTest.testBasicUploadOCA\DAV\Tests\unit\Connector\Sabre\RequestTest\PartFileInRootUploadTest.testBasicUpload
-- | --
```

**Comment by mrow4a** [comment]
Unit tests are not adjusted because they don't pass PATH_INFO for $SERVER

**Comment by mrow4a** [comment]
@jvillafanez @PVince81 @DeepDiver1975 review here?

File: apps/dav/lib/Upload/FutureFile.php [hunk 1]
```diff
@@ -40,6 +40,23 @@ class FutureFile implements \Sabre\DAV\IFile {
...
 
+	static public function getFutureFileName() {
+		return '.file';
+	}
+
+	static public function isFutureFile() {
+		$davUploadsTarget = '/dav/uploads';
+
+		// Check if pathinfo starts with dav uploads target and basename is future file basename
+		if (isset($_SERVER['PATH_INFO'])
```
Comment (PVince81) [inline apps/dav/lib/Upload/FutureFile.php]: any way to do this without global variables + static functions ? (pretty sure @DeepDiver1975 would not be happy either)

I do see that it is likely rather tricky to inject all this down to the "File" class.
@DeepDiver1975 should we pass on this one ?

Comment (mrow4a) [inline apps/dav/lib/Upload/FutureFile.php]: Honestly, it was just the suggestion, but wanted to create a specific class which might hold all these values.

Comment (mrow4a) [inline apps/dav/lib/Upload/FutureFile.php]: About the global, not sure if request holds this value alone.

**Review by PVince81 — APPROVED** [review]
:+1: fine by me

**Comment by PVince81** [comment]
@mrow4a please backport to stable10

**Pull request merged**